internal/service/advancedcluster/resource_advanced_cluster.go

```diff
@@ -667,15 +625,31 @@ func resourceRead(ctx context.Context, d *schema.ResourceData, meta any) diag.Di
 	return nil
 }
 
+// GetReplicationSpecAttributesFromOldAPI returns the id and num shard values of replication specs coming from old API. This is used to populate replication_specs.*.id and replication_specs.*.num_shard attributes for old sharding confirgurations.
+// In the old API, each replications spec has a 1:1 relation with each zone, so ids and num shards are stored in a struct oldShardConfigMeta and are returned in a map from zoneName to oldShardConfigMeta.
+func GetReplicationSpecAttributesFromOldAPI(ctx context.Context, projectID, clusterName string, client20240530 admin20240530.ClustersApi) (map[string]OldShardConfigMeta, error) {
+	clusterOldAPI, _, err := client20240530.GetCluster(ctx, projectID, clusterName).Execute()
+	if err != nil {
+		readErrorMsg := "error reading advanced cluster with 2023-02-01 API (%s): %s"
+		return nil, fmt.Errorf(readErrorMsg, clusterName, err)
+	}
+	specs := clusterOldAPI.GetReplicationSpecs()
+	result := make(map[string]OldShardConfigMeta, len(specs))
+	for _, spec := range specs {
+		result[spec.GetZoneName()] = OldShardConfigMeta{spec.GetId(), spec.GetNumShards()}
+	}
+	return result, nil
+}
+
 // getReplicationSpecIDsFromOldAPI returns the id values of replication specs coming from old API. This is used to populate old replication_specs.*.id attribute avoiding breaking changes.
 // In the old API each replications spec has a 1:1 relation with each zone, so ids are returned in a map from zoneName to id.
 func getReplicationSpecIDsFromOldAPI(ctx context.Context, projectID, clusterName string, connV220240530 *admin20240530.APIClient) (map[string]string, error) {
```

Comment:
I see this is still used in the data source, but probably we can re-use the new `GetReplicationSpecAttributesFromOldAPI` instead?

Reply:
data source refactor (and clean up) will be done as part of CLOUDP-287294 to keep the PR as small as possible